cmd/intercept/intercept.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"sda-pipeline/internal/broker"
+	"sda-pipeline/internal/config"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+func main() {
+	conf, err := config.NewConfig("intercept")
+	if err != nil {
+		log.Fatal(err)
+	}
+	mq, err := broker.NewMQ(conf.Broker)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer mq.Channel.Close()
+	defer mq.Connection.Close()
+
+	forever := make(chan bool)
+
+	log.Info("starting intercept service")
+
+	go func() {
+		messages, err := broker.GetMessages(mq, conf.Broker.Queue)
+		if err != nil {
+			log.Fatal(err)
+		}
+		for delivered := range messages {
+			log.Debugf("received a message: %s", delivered.Body)
+			msgType, res, err := validateJSON(delivered.Body)
+			if res == nil {
+				continue
+			}
+			if err != nil {
+				log.Error(err)
+				// publish MQ error
+				continue
+			}
+			if !res.Valid() {
+				log.Error(res.Errors())
+				// publish MQ error
+				continue
+			}
+
+			var routingKey string
+
+			switch msgType {
+			case "accession":
+				routingKey = "stableIDs"
+			case "cancel":
+				routingKey = ""
+				continue
+			case "ingest":
+				routingKey = "files"
+			case "mapping":
+				routingKey = "mappings"
+			}
+
+			if err := broker.SendMessage(mq, delivered.CorrelationId, conf.Broker.Exchange, routingKey, conf.Broker.Durable, delivered.Body); err != nil {
+				// TODO fix resend mechanism
+				log.Errorln("We need to fix this resend stuff ...")
+			}
+			if err := delivered.Ack(false); err != nil {
+				log.Errorf("failed to ack message for reason: %v", err)
+			}
+		}
+	}()
+
+	<-forever
+}
+
+func validateJSON(body []byte) (string, *gojsonschema.Result, error) {
```

Comment:
I'm unsure about how much heavy lifting the NewReferenceLoader does. Fine for now, but if we ever see visible resources usage from this, we might want to load these just once and reuse.

Reply:
I don't foresee a high load in this place, we won't get so many messages so frequently. It's gonna be okay 👌